2015-Fall/oPolo/Calculator/Calculator/CalculatorBrain.swift

```diff
@@ -0,0 +1,114 @@
+                case .BinaryOperation(let symbol, _):
+                    return symbol
+                }
+            }
+        }
+    }
+    
+    private var opStack = [Op]() {
+        didSet {
+            //Fires a notification when a new operand/operator/constant/etc is added to the stack.
+            NSNotificationCenter.defaultCenter().postNotificationName("QueuedCalculationsChanged", object: nil)
+        }
+    }
+    
+    
+    private var knownOps = [String:Op]()
+    
+    init() {
+        knownOps["π"] = Op.Constant("π", M_PI)
+        knownOps["×"] = Op.BinaryOperation("×", *)
+        knownOps["÷"] = Op.BinaryOperation("÷") { $1 / $0}
+        knownOps["+"] = Op.BinaryOperation("+", +)
+        knownOps["−"] = Op.BinaryOperation("−") { $1 - $0}
+        knownOps["√"] = Op.UnaryOperation("√", sqrt)
+        knownOps["sin"] = Op.UnaryOperation("sin", sin)
+        knownOps["cos"] = Op.UnaryOperation("cos", cos)
+    }
+    
+    /*
+    * This function returns a string representation all the operations, operands, etc that the 
+    * calculator has currently queued up.
+    */
+    func queuedOperations() -> String {
+        //Note: While this function is currently not doing it, 
+        //this is one obvious place for other objects to get a nicely
+        //formatted "math" string with parentheses where due, etc.
+        //e.g. [2, 2, 3, +, x] becomes (2+3)*2
+        return "\(opStack)"
+    }
+    
+    private func evaluate(ops: [Op]) -> (result: Double?, remainingOps: [Op]) {
+        if !ops.isEmpty {
+            var remainingOps = ops
+            let op = remainingOps.removeLast()
+            switch op {
+            case .Operand(let operand):
+                return (operand, remainingOps)
+            case .Constant(_, let constantValue):
+                return (constantValue, remainingOps)
+            case .UnaryOperation(_, let operation):
+                let operandEvaluation = evaluate(remainingOps)
+                if let operand = operandEvaluation.result {
+                    return (operation(operand), operandEvaluation.remainingOps)
+                }
+            case .BinaryOperation(_, let operation):
+                let op1Evaluation = evaluate(remainingOps)
+                if let operand1 = op1Evaluation.result {
+                    let op2Evaluation = evaluate(op1Evaluation.remainingOps)
+                    if let operand2 = op2Evaluation.result {
+                        return (operation(operand1, operand2), op2Evaluation.remainingOps)
+                    }
+                }
+            }
+        }
+        return (nil, ops)
+    }
+    
+    private func evaluate() -> Double? {
+        let (result, remainder) = evaluate(opStack)
+        return result
+    }
+    
+    func pushOperand(operand: Double) -> Double? {
+        opStack.append(Op.Operand(operand))
+        return evaluate()
+    }
+    
+    func resetCalculator() {
+        opStack = [Op]()
```

Comment:
Agreed. It is a good general OOP question that I think we should talk a bit about next time we meet.